13_sandboxes/jupyter_sandbox.py

```diff
@@ -0,0 +1,69 @@
+# ---
+# cmd: ["python", "13_sandboxes/jupyter_sandbox.py"]
+# tags: ["use-case-sandboxed-code-execution"]
+# ---
+
+# # Run a Jupyter notebook in a Modal Sandbox
+#
+# This example demonstrates how to run a Jupyter notebook in a Modal [Sandbox](/docs/guide/sandbox).
+
+import secrets
+
+import modal
+
+# All Sandboxes are associated with an App. We look up our app by name, creating it if it doesn't exist.
+
+app = modal.App.lookup("example-jupyter", create_if_missing=True)
+
+# We define a custom Docker image that has Jupyter and some other dependencies installed.
+# Using a custom image allows us to avoid re-installing these deps on every Sandbox startup.
+
+image = (
+    modal.Image.debian_slim()
+    .pip_install("jupyter~=1.1.0", "bing-image-downloader~=1.1.2")
+    .pip_install("pandas~=2.2.3")  # Any other deps
+)
+
+# Since we'll be exposing a Jupyter server, we'll need to create a password.
+# We'll use secrets to store the token in a Modal Secret.
+
+token = secrets.token_urlsafe(13)
+token_secret = modal.Secret.from_dict({"JUPYTER_TOKEN": token})
+
+# Now, we can start our sandbox. Note our use of the `encrypted_ports` argument, which
+# allows us to securely expose the Jupyter server to the public internet. We use
+# `modal.enable_output()` to print the Sandbox's image build logs to the console.
+
+JUPYTER_PORT = 8888
+
+with modal.enable_output():
+    sandbox = modal.Sandbox.create(
+        "jupyter",
+        "notebook",
+        "--no-browser",
+        "--allow-root",
+        "--ip=0.0.0.0",
+        f"--port={JUPYTER_PORT}",
+        "--NotebookApp.allow_origin='*'",
+        "--NotebookApp.allow_remote_access=1",
+        encrypted_ports=[JUPYTER_PORT],
+        secrets=[token_secret],
+        timeout=60 * 60,  # 1 hour
+        image=image,
+        app=app,
+    )
+
+print(f"Sandbox ID: {sandbox.object_id}")
+
+# Finally, we can print out a URL that we can use to connect to our Jupyter server.
+# Note that we have to call [`Sandbox.tunnels`](/docs/reference/modal.Sandbox#tunnels) to get the
+# URL because the Sandbox is not publicly accessible until we do so.
+
+tunnel = sandbox.tunnels()[JUPYTER_PORT]
+url = f"{tunnel.url}/?token={token}"
+print(f"Jupyter notebook is running at: {url}")
+
+# You can now open this URL in your browser to access the Jupyter notebook! When you're done,
```

Comment:
Would love a line of code or two to test this -- a quick and cutty manual urllib request to a Jupyter API endpoint that demonstrates the server is live.

More for testing than for users.

Reply:
done